cirq/circuits/expectation_value.py

```diff
@@ -0,0 +1,376 @@
+
+            # caculates expectation
+
+            q1 = q_dict[list(term.qubits())[0]]
+            if len(term) == 2:
+                q2 = q_dict[list(term.qubits())[1]]
+            else:
+                q2 = q1
+
+            x1 = -2 * int(results_dict[q1][rep_term][0]) + 1
+            x2 = -2 * int(results_dict[q2][rep_term][0]) + 1
+            print(x1,x2)
+
+            expectation += x1 * coef * x2
+
+        expectation_samples.append(expectation)
+
+    expectation_mean = np.mean(expectation_samples) + identity_coeficient
+
+    return expectation_mean
+
+
+def expectation_value(circuit: Circuit,
+                      operator: Dict[PauliString, float]
+                      ):
+    """
+    Calculates the expactation value of cost function (operator)
+    for a circuit run through a simulator by taking an inner product with
+    the wavefunction
+
+    Args:
+        circuit: circ.Circuit type which prepares the state we are interested
+                 in calculating the expectation value of
+
+        operator: Operator input as a dictionary where keys are
+                    cirq.ops.PauliString type and values are coefficients
+                  example:
+                  paulistring = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.X, qubits[1]: (cirq.Pauli.Z)})
+                  paulistring2 = cirq.PauliString(qubit_pauli_map=
+                    {qubits[0]:cirq.Pauli.Z, qubits[2]: (cirq.Pauli.Y)})
+                  operator = {paulistring: 1.234, paulistring2: -5.678}
+
+    Returns:
+        expectation value of operator argument in the state prepared by
+        circuit.
+
+    """
+
+    qubits = list(circuit.all_qubits())
+    sim = XmonSimulator()
+
+    # ensures no measurements are performed:
+    if any(MeasurementGate.is_measurement(op) for op
+           in circuit.all_operations()):
+        raise ValueError("Circuit has measurements.")
+
+    # number of qubits
+    n_qubits = len(qubits)
+
+    # setup the bit strings for each final state:
+    # bits = [bin(i)[2:].zfill(n_qubits) for i in range(2**n_qubits)]
+
+    # runs circuit:
+    results = sim.simulate(circuit)
+
+    # setup the probabilities for each state
+    final_state = results.final_state
+
+    expectation = 0
+
+    # Arrays to help in computation
+
+    pauli_x = np.array([[0, 1], [1, 0]])
+    pauli_y = np.array([[0, -1j], [1j, 0]])
+    pauli_z = np.array([[1, 0], [0, -1]])
+    identity = np.identity(2)
+
+    identity_coeficient = 0
+    for term, coef in operator.items():
+        # reset operator
+        full_op_list = [identity for i in range(n_qubits)]
+
+        # skip over identity terms
+        if len(term) == 0:
+            identity_coeficient = coef
+            continue
+
+        # contruct correct operator
+        for qubit, pauli in term.items():
+
+            q_idx = qubits.index(qubit)
+
+            if pauli == Pauli.X:
```

Comment:
Use a dictionary `pauli_to_full_op` defined outside the for loop instead of a multi-if.